File: C1.UWP.Input/VB/InputSamples/Data/DataProvider.vb
Imports System.Xml.Serialization
Imports System.Threading.Tasks
Imports System.Text
Imports System.Reflection
Imports System.Linq
Imports System.Collections.ObjectModel
Imports System.Collections.Generic
Imports System
Imports C1.Xaml.Input
Imports C1.Xaml

Public Class DataProvider
    Private incomeCollection As C1CollectionView
    Private expenseCollection As C1CollectionView

    Private _distributionData As DistributionData
    Private finnanceData As FinnanceData
    Private addressBook As AddressBook

    Shared provider As DataProvider

    Public Property FilterFamilyMembers() As List(Of String)
    Public Property FilterAccountTypes() As List(Of AccountType)
    Public Property FilterIncomeTypes() As List(Of IncomeType)
    Public Property FilterExpenseTypes() As List(Of ExpenseType)

    Public ReadOnly Property DistributionData() As DistributionData
        Get
            Return _distributionData
        End Get
    End Property

    Public ReadOnly Property Incomes() As C1CollectionView
        Get
            Return incomeCollection
        End Get
    End Property
    Public ReadOnly Property Expenses() As C1CollectionView
        Get
            Return expenseCollection
        End Get
    End Property

    Public ReadOnly Property AccountTypes() As Dictionary(Of String, AccountType)
        Get
            Dim dictionary As New Dictionary(Of String, AccountType)()
            dictionary(Strings.Account_Cash) = AccountType.Cash
            dictionary(Strings.Account_CreditCard) = AccountType.CreditCard
            dictionary(Strings.Account_BankAccount) = AccountType.BankAccount
            dictionary(Strings.Account_OnlinePayment) = AccountType.OnlinePayment
            Return dictionary
        End Get
    End Property
    Public ReadOnly Property IncomeTypes() As Dictionary(Of String, IncomeType)
        Get
            Dim dictionary As New Dictionary(Of String, IncomeType)()
            dictionary(Strings.IncomeType_Salary) = IncomeType.Salary
            dictionary(Strings.IncomeType_Stock) = IncomeType.Stock
            dictionary(Strings.IncomeType_InvestmentReturn) = IncomeType.Investment
            dictionary(Strings.IncomeType_Lottery) = IncomeType.Lottery
            dictionary(Strings.IncomeType_Others) = IncomeType.Others
            Return dictionary
        End Get
    End Property
    Public ReadOnly Property ExpenseTypes() As Dictionary(Of String, ExpenseType)
        Get
            Dim dictionary As New Dictionary(Of String, ExpenseType)()
            dictionary(Strings.ExpenseType_Tax) = ExpenseType.Tax
            dictionary(Strings.ExpenseType_Traffic) = ExpenseType.Traffic
            dictionary(Strings.ExpenseType_Shopping) = ExpenseType.Shopping
            dictionary(Strings.ExpenseType_Entertainment) = ExpenseType.Entertainment
            dictionary(Strings.ExpenseType_Education) = ExpenseType.Education
            dictionary(Strings.ExpenseType_Health) = ExpenseType.Health
            dictionary(Strings.ExpenseType_Others) = ExpenseType.Others
            Return dictionary
        End Get
    End Property

    Public ReadOnly Property FamilyMembers() As List(Of String)
        Get
            Dim list As New List(Of String)()
            list.Add("Richard")
            list.Add("Violet")
            list.Add("Olivia")
            list.Add("Lucas")
            Return list
        End Get
    End Property

    Public ReadOnly Property MailList() As List(Of Mail)
        Get
            Return addressBook.Mails
        End Get
    End Property
    Shared Sub New()
    End Sub

    Public Shared Function GetProvider() As DataProvider
        If provider Is Nothing Then
            provider = New DataProvider()
        End If
        Return provider
    End Function

    Public Sub InstallData()
        Dim assembly As Assembly = Assembly.Load(New AssemblyName("InputSamplesLib"))
        Dim mapDataStream As Stream = assembly.GetManifestResourceStream("InputSamples.MapData.xml")
        _distributionData = CType(New XmlSerializer(GetType(DistributionData)).Deserialize(mapDataStream), DistributionData)
        Dim finnanceDataStream As Stream = assembly.GetManifestResourceStream("InputSamples.FinanceData.xml")
        finnanceData = CType(New XmlSerializer(GetType(FinnanceData)).Deserialize(finnanceDataStream), FinnanceData)
        Dim incomesCollcetion As New ObservableCollection(Of Income)(finnanceData.Incomes)
        Dim mailDataStream As Stream = assembly.GetManifestResourceStream("InputSamples.MailData.xml")
        addressBook = CType(New XmlSerializer(GetType(AddressBook)).Deserialize(mailDataStream), AddressBook)
        incomeCollection = New C1CollectionView(finnanceData.Incomes)
        incomeCollection.GroupDescriptions.Add(New PropertyGroupDescription("Name"))
        incomeCollection.GroupDescriptions.Add(New PropertyGroupDescription("AccountType"))
        incomeCollection.GroupDescriptions.Add(New PropertyGroupDescription("Type"))
        expenseCollection = New C1CollectionView(finnanceData.Expenses)
        expenseCollection.GroupDescriptions.Add(New PropertyGroupDescription("Name"))
        expenseCollection.GroupDescriptions.Add(New PropertyGroupDescription("AccountType"))
        expenseCollection.GroupDescriptions.Add(New PropertyGroupDescription("Type"))
        FilterFamilyMembers = New List(Of String)()
        FilterAccountTypes = New List(Of AccountType)()
        FilterIncomeTypes = New List(Of IncomeType)()
        FilterExpenseTypes = New List(Of ExpenseType)()
        incomeCollection.Filter = New Predicate(Of Object)(AddressOf FilterCompatibleIncomeItems)
        expenseCollection.Filter = New Predicate(Of Object)(AddressOf FilterCompatibleExpenseItems)
    End Sub

    Private Function FilterCompatibleIncomeItems(incomeItem As Object) As Boolean
        Dim current As Income = TryCast(incomeItem, Income)
        If current Is Nothing Then
            Return False
        End If
        Dim result As Boolean = FilterFamilyMembers.Contains(current.Name) AndAlso FilterAccountTypes.Contains(current.AccountType) AndAlso FilterIncomeTypes.Contains(current.Type)
        Return result
    End Function
    Private Function FilterCompatibleExpenseItems(expenseItem As Object) As Boolean
        Dim current As Expense = TryCast(expenseItem, Expense)
        If current Is Nothing Then
            Return False
        End If
        Dim result As Boolean = FilterFamilyMembers.Contains(current.Name) AndAlso FilterAccountTypes.Contains(current.AccountType) AndAlso FilterExpenseTypes.Contains(current.Type)
        Return result
    End Function

    Public Sub UpdateData(Of T)(addItems As List(Of T), removeItems As List(Of T), source As List(Of T), filterType As DataFilterType)
        For Each item As T In addItems
            If Not source.Contains(item) Then
                source.Add(item)
            End If
        Next
        For Each item As T In removeItems
            If source.Contains(item) Then
                source.Remove(item)
            End If
        Next
        If filterType <> DataFilterType.Income Then
            expenseCollection.Filter = Nothing
            expenseCollection.Filter = New Predicate(Of Object)(AddressOf FilterCompatibleExpenseItems)
        End If
        If filterType <> DataFilterType.Expense Then
            incomeCollection.Filter = Nothing
            incomeCollection.Filter = New Predicate(Of Object)(AddressOf FilterCompatibleIncomeItems)
        End If
    End Sub
End Class

Public Enum DataFilterType
    Name
    Account
    Income
    Expense
End Enum

File: C1.UWP.Input/CS/InputSamples/Data/DataProvider.cs
using C1.Xaml;
using C1.Xaml.Input;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;

namespace InputSamples
{
    public class DataProvider
    {
        private C1CollectionView incomeCollection;
        private C1CollectionView expenseCollection;

        private DistributionData distributionData;
        private FinnanceData finnanceData;
        private AddressBook addressBook;

        static DataProvider provider;

        public List<string> FilterFamilyMembers { get; private set; }
        public List<AccountType> FilterAccountTypes { get; private set; }
        public List<IncomeType> FilterIncomeTypes { get; private set; }
        public List<ExpenseType> FilterExpenseTypes { get; private set; }

        public DistributionData DistributionData { get { return distributionData; } }

        public C1CollectionView Incomes { get { return incomeCollection; } }
        public C1CollectionView Expenses { get { return expenseCollection; } }

        public Dictionary<string, AccountType> AccountTypes
        {
            get
            {
                Dictionary<string, AccountType> dictionary = new Dictionary<string, AccountType>();
                dictionary[Strings.Account_Cash] = AccountType.Cash;
                dictionary[Strings.Account_CreditCard] = AccountType.CreditCard;
                dictionary[Strings.Account_BankAccount] = AccountType.BankAccount;
                dictionary[Strings.Account_OnlinePayment] = AccountType.OnlinePayment;
                return dictionary;
            }
        }
        public Dictionary<string, IncomeType> IncomeTypes
        {
            get
            {
                Dictionary<string, IncomeType> dictionary = new Dictionary<string, IncomeType>();
                dictionary[Strings.IncomeType_Salary] = IncomeType.Salary;
                dictionary[Strings.IncomeType_Stock] = IncomeType.Stock;
                dictionary[Strings.IncomeType_InvestmentReturn] = IncomeType.Investment;
                dictionary[Strings.IncomeType_Lottery] = IncomeType.Lottery;
                dictionary[Strings.IncomeType_Others] = IncomeType.Others;
                return dictionary;
            }
        }
        public Dictionary<string, ExpenseType> ExpenseTypes
        {
            get
            {
                Dictionary<string, ExpenseType> dictionary = new Dictionary<string, ExpenseType>();
                dictionary[Strings.ExpenseType_Tax] = ExpenseType.Tax;
                dictionary[Strings.ExpenseType_Traffic] = ExpenseType.Traffic;
                dictionary[Strings.ExpenseType_Shopping] = ExpenseType.Shopping;
                dictionary[Strings.ExpenseType_Entertainment] = ExpenseType.Entertainment;
                dictionary[Strings.ExpenseType_Education] = ExpenseType.Education;
                dictionary[Strings.ExpenseType_Health] = ExpenseType.Health;
                dictionary[Strings.ExpenseType_Others] = ExpenseType.Others;
                return dictionary;
            }
        }

        public List<string> FamilyMembers
        {
            get
            {
                List<string> list = new List<string>();
                list.Add("Richard");
                list.Add("Violet");
                list.Add("Olivia");
                list.Add("Lucas");
                return list;
            }
        }

        public List<Mail> MailList { get { return addressBook.Mails; } }

        static DataProvider() { }

        public static DataProvider GetProvider()
        {
            if (provider == null)
            {
                provider = new DataProvider();
                provider.InstallData();
            }
            return provider;
        }

        public void InstallData()
        {
            Assembly assembly = Assembly.Load(new AssemblyName("InputSamplesLib"));
            var mapDataStream = assembly.GetManifestResourceStream("InputSamples.Resources.MapData.xml");
            distributionData = (DistributionData)new XmlSerializer(typeof(DistributionData)).Deserialize(mapDataStream);
            var finnanceDataStream = assembly.GetManifestResourceStream("InputSamples.Resources.FinanceData.xml");
            finnanceData = (FinnanceData)new XmlSerializer(typeof(FinnanceData)).Deserialize(finnanceDataStream);
            ObservableCollection<Income> incomesCollcetion = new ObservableCollection<Income>(finnanceData.Incomes);
            var mailDataStream = assembly.GetManifestResourceStream("InputSamples.Resources.MailData.xml");
            addressBook = (AddressBook)new XmlSerializer(typeof(AddressBook)).Deserialize(mailDataStream);
            incomeCollection = new C1CollectionView(finnanceData.Incomes);
            incomeCollection.GroupDescriptions.Add(new PropertyGroupDescription("Name"));
            incomeCollection.GroupDescriptions.Add(new PropertyGroupDescription("AccountType"));
            incomeCollection.GroupDescriptions.Add(new PropertyGroupDescription("Type"));
            expenseCollection = new C1CollectionView(finnanceData.Expenses);
            expenseCollection.GroupDescriptions.Add(new PropertyGroupDescription("Name"));
            expenseCollection.GroupDescriptions.Add(new PropertyGroupDescription("AccountType"));
            expenseCollection.GroupDescriptions.Add(new PropertyGroupDescription("Type"));
            FilterFamilyMembers = new List<string>();
            FilterAccountTypes = new List<AccountType>();
            FilterIncomeTypes = new List<IncomeType>();
            FilterExpenseTypes = new List<ExpenseType>();
            incomeCollection.Filter = new Predicate<object>(FilterCompatibleIncomeItems);
            expenseCollection.Filter = new Predicate<object>(FilterCompatibleExpenseItems);
        }

        private bool FilterCompatibleIncomeItems(object incomeItem)
        {
            var current = incomeItem as Income;
            if (current == null)
                return false;
            bool result = FilterFamilyMembers.Contains(current.Name) && FilterAccountTypes.Contains(current.AccountType) && FilterIncomeTypes.Contains(current.Type);
            return result;
        }
        private bool FilterCompatibleExpenseItems(object expenseItem)
        {
            var current = expenseItem as Expense;
            if (current == null)
                return false;
            bool result = FilterFamilyMembers.Contains(current.Name) && FilterAccountTypes.Contains(current.AccountType) && FilterExpenseTypes.Contains(current.Type);
            return result;
        }

        public void UpdateData<T>(List<T> addItems, List<T> removeItems, List<T> source, DataFilterType filterType)
        {
            foreach (T item in addItems)
            {
                if (!source.Contains(item))
                    source.Add(item);
            }
            foreach (T item in removeItems)
            {
                if (source.Contains(item))
                    source.Remove(item);
            }
            if (filterType != DataFilterType.Income)
            {
                expenseCollection.Filter = null;
                expenseCollection.Filter = new Predicate<object>(FilterCompatibleExpenseItems);
            }
            if (filterType != DataFilterType.Expense)
            {
                incomeCollection.Filter = null;
                incomeCollection.Filter = new Predicate<object>(FilterCompatibleIncomeItems);
            }
        }
    }

    public enum DataFilterType
    {
        Name,
        Account,
        Income,
        Expense
    }
}
